src/hmem_ipc_cache.c

```diff
@@ -0,0 +1,231 @@
+{
+	/* 
+	 * The ipc monitor should be general for any iface.
+	 * Start from FI_HMEM_CUDA and we can expand it later.
+	 */
+	struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
+		[FI_HMEM_CUDA] = ipc_monitor,
+	};
+	struct util_domain *util_domain;
+	int ret;
+	util_domain = container_of(domain, struct util_domain, domain_fid);
+	*cache = (struct ofi_mr_cache *)calloc(1, sizeof(struct ofi_mr_cache));
+	if (!*cache)
+		return -FI_ENOMEM;
+	(*cache)->add_region = ipc_cache_add_region;
+	(*cache)->delete_region = ipc_cache_delete_region;
+	ret = ofi_mr_cache_init(util_domain, memory_monitors,
+				*cache);
+	if (ret) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"ipc cache init failed: %s\n",
+		        fi_strerror(ret));
+		free(*cache);
+		*cache = NULL;
+		return -ret;
+	}
+
+	FI_INFO(&core_prov, FI_LOG_CORE,
+		"ipc cache enabled, max_cnt: %zu max_size: %zu\n",
+		cache_params.max_cnt, cache_params.max_size);
+	return 0;
+}
+
+/**
+ * @brief Destroy the ipc cache
+ * 
+ * @param cache the ipc cache
+ */
+void ofi_ipc_cache_destroy(struct ofi_mr_cache *cache)
+{
+	ofi_mr_cache_cleanup(cache);
+	free(cache);
+	cache = NULL;
+}
+
+/**
+ * @brief Given ipc_info (with ipc_handle and the iov of the device allocation),
+ * assign the mapped_addr the mapped address of the ipc_handle.
+ * Each (ipc_handle, mapped_addr) pair is stored in ofi_mr_entry.info.ipc_info as
+ * part of each mr entry.
+ * In a cache hit, the mapped_addr is retrieved from the matched mr entry. Otherwise,
+ * the mapped_addr is obtained by opening the ipc handle.
+ * 
+ * @param[in] cache the ipc cache
+ * @param[in] ipc_info the information of the ipc to be mapped.
+ * @param[out] mapped_addr the mapped address for the ipc handle.
+ * @param[out] mr_entry the matched mr_entry of the ipc_info and mapped_addr.
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_search(struct ofi_mr_cache *cache, struct smr_ipc_info *ipc_info,
+			  void **mapped_addr, struct ofi_mr_entry **mr_entry)
+{
+	struct fi_mr_attr attr;
+	struct iovec iov;
+	struct ofi_mr_entry *entry;
+	int ret;
+	ofi_genlock_lock(&cache->domain->lock);
+	iov.iov_base = (void *)ipc_info->base_address;
+	iov.iov_len = ipc_info->base_length;
+	attr.mr_iov = &iov;
+	attr.iface = ipc_info->iface;
+	attr.iov_count = 1;
+
+	ret = ofi_mr_cache_search(cache, &attr, &entry);
+	if (OFI_UNLIKELY(ret)) {
+		ofi_genlock_unlock(&cache->domain->lock);
+		return ret;
+	}
+
+	if (!entry->info.ipc_mapped_addr) {
+		/*
+		 * This indicates that the mr entry is inserted in the cache the first time 
+		 * thus does not have the ipc_mapped_addr assigned in the info.
+		 */
+		memcpy((void **)&entry->info.ipc_handle, (void **) &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache opens new ipc handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		ret = ipc_cache_open_handle(cache, entry);
```

Comment:
Don't you still need to handle the -FI_EALREADY error by calling `ofi_mr_flush` and retry?

Reply:
Currently, we call `ofi_mr_cache_flush` when hitting `-FI_ELAREADY`. If we can do `ofi_mr_cache_flush` every time before opening the handle, we should not need to handle `FI_ELAREADY` explicitly then.

Reply:
I see. It is handled in add_regions.

Reply:
Also, @wckzhang proposed an idea of making the length of the mr_region to be only 1 for ipc cach. By this design, even if the receiver got a new IPC handle with the same base address but larger allocation, the ofi_mr_cache_search will regard it as a cache hit, then `ofi_ipc_cache_search` will invalidate the old IPC handle by closing it and open the new ipc handle. So there should be no `FI_EALREADY ` for sure.

I cannot think of a corner case that can make this idea fail.

Reply:
@wzamazon 
> I see. It is handled in add_regions.

No, now we have to make `cache->add_region` a no-op since we keep the `ofi_mr_cache_search` API not changed. The handle open are fully managed inside `ofi_ipc_cache_search`, by calling `ipc_cache_open_handle()`